BlueDBOnDisk/src/main/java/io/bluedb/disk/recovery/RecoveryManager.java

```diff
@@ -8,44 +8,97 @@
 	protected static String SUFFIX = ".chg";
+	private static long FREQUENCY_OF_COMPLETED_CLEANUP = 10 * 60 * 1000; // ten minutes
+	private static long COMPLETED_RETENTION = 60 * 60 * 1000; // one hour
 
 	private final BlueCollectionOnDisk<T> collection;
 	private final Path recoveryPath;
+	private final Path pendingFolderPath;
+	private final Path historyFolderPath;
 	private final FileManager fileManager;
+	private long lastLogCleanup = 0;
 	
 	public RecoveryManager(BlueCollectionOnDisk<T> collection, FileManager fileManager, BlueSerializer serializer) {
 		this.collection = collection;
 		this.fileManager = fileManager;
-		this.recoveryPath = Paths.get(collection.getPath().toString(), SUBFOLDER);
+		this.recoveryPath = Paths.get(collection.getPath().toString(), RECOVERY_FOLDER);
+		this.pendingFolderPath = Paths.get(recoveryPath.toString(), PENDING_SUBFOLDER);
+		this.historyFolderPath = Paths.get(recoveryPath.toString(), HISTORY_SUBFOLDER);
 	}
 
 	public void saveChange(Recoverable<?> change) throws BlueDbException {
 		String filename = getFileName(change);
-		Path path = Paths.get(recoveryPath.toString(), filename);
-		fileManager.saveObject(path, change);
+		Path pendingPath = Paths.get(pendingFolderPath.toString(), filename);
+		Path historyPath = Paths.get(historyFolderPath.toString(), filename);
+		fileManager.saveObject(pendingPath, change);
+		fileManager.saveObject(historyPath, change);
+	}
+
+	public Path getPendingFolderPath() {
+		return pendingFolderPath;
 	}
 
 	public void removeChange(Recoverable<?> change) throws BlueDbException {
 		String filename = getFileName(change);
-		Path path = Paths.get(recoveryPath.toString(), filename);
+		Path path = Paths.get(pendingFolderPath.toString(), filename);
 		File file = new File(path.toString());
 		file.delete();
+		if (isTimeForHistoryCleanup()) {
+			cleanupHistory();  // TODO run in a different thread?
+		}
 	}
 
-	public static String getFileName(Recoverable<?> change) {
-		return  String.valueOf(change.getTimeCreated()) + SUFFIX;
+	protected boolean isTimeForHistoryCleanup() {
+		long timeSinceLastCleanup = System.currentTimeMillis() - lastLogCleanup;
+		return timeSinceLastCleanup > FREQUENCY_OF_COMPLETED_CLEANUP;
+	}
+
+	protected void cleanupHistory() {
+		// TODO check if restore is pending
+		List<File> historicChangeFiles = FileManager.getFolderContents(historyFolderPath, SUFFIX);
+		List<File> filesOldEnoughToDelete = Blutils.filter(historicChangeFiles, (f) -> isChangeFileOlderThan(f, COMPLETED_RETENTION));
+		filesOldEnoughToDelete.forEach((File f) -> f.delete());
+		lastLogCleanup = System.currentTimeMillis();
+	}
+
+	public List<File> getChangeHistory(long backupStartTime, long backupEndTime) throws BlueDbException {
+		List<File> files = FileManager.getFolderContents(historyFolderPath, SUFFIX);
+		if (files.isEmpty()) {
+			return files;
+		}
+		Blutils.sortByMappedvalue(files, (File f) -> extractTimestamp(f) );

```

Comment:
You should camel-cap the whole word. `sortByMappedValue`. 

Your `value` is lower case right now.

Reply:
done